### PR TITLE
fix: composer not cleared when sending message

### PR DESCRIPTION
**Pull request by hannojg**
### Details

([Slack thread](https://expensify.slack.com/archives/C05R2V5GM1S/p1718186137117569))

This PR fixes an issue where the composer's text input wouldn't show the value the user expects. The following use case now works:

- Spam the 'a' key, typing 'aaaaaa...'
- Press send
- Spam the 'b' key, typing 'bbbbbb...'
- Expect that the composer shows first 'aaaaaa', then clears it, then it shows 'bbbbbb'

In react native the text input sends the whole text thats currently in the text input in the onChange/onChangeText event callbacks. However, we maintain the "source of truth" for our text input in JavaScript in our `value` react state variable.
So it can happen that we set our JS `value` to `""` to clear the input, but the native side is ignoring that event (as the JS thread is lagging behind in events), and sends us a new update with the current native input (which would be "aaaaaabbbbbb" by then).

With the approach in this PR, we only apply the difference from the native side to our JS value. E.g. if the user pressed another "a", instead of setting the whole native text to our js `value` we just append the `a` that was added.

This way the native side might still ignore our clear event (setting the value to `""`), but when the user presses another character like `"b"` it will be appended to our empty value `"" + "b"`, and when the native side then processes this update, it shows correctly `"b"` instead of `"aaaaaab"`.

For this to work we need to receive the range in which the changes to the new active text have been made. This requires changes in `react-native` and `@expensify/react-native-live-markdown`. Two patches have been added for which we have PRs here:

- https://github.com/facebook/react-native/pull/45248
- https://github.com/Expensify/react-native-live-markdown/pull/412

This PR bumps rn-live-markdown to include the changes necessary for this PR. Before merging this PR make sure to merge this one first:

- https://github.com/Expensify/App/pull/44958

#### TODO:

- [ ] ~~Investigate potential re-render loop:~~ (Shouldn't be part of this PR, this PR didn't;t introduce it)
	- Enter some text
	- Reload the app
	- Remove the draft text
	- At this point I get a "Maximum call depth warning" - not sure if introduced by this PR, but there _might_ be something wrong
- [x] Bug: Write a message with a mention, send the message, press on the mention, go back, the text will reappear in the composer 
- [x] Make changes to react-native-web
- [x] Test on web
	- [x] Test writing a comment in one tab, check that the other tab updates
- [x] Fix web bugs: before, count, start are often incorrect for the different types of events
- [x] Fix web bug where multi-character glyphs are not deleted correctly
- [x] Fix Todos / create issues for it / revert some of the earlier code cleanups that just bloat the changes of this PR
- [x] Raise discussion in RN ([link](https://github.com/facebook/react-native/issues/37722))
- [x] Make PR in RN ([link](https://github.com/facebook/react-native/pull/45248))
- [x] Make a PR in `@expensify/react-native-live-markdown`

### Fixed Issues

$https://github.com/Expensify/App/issues/37896
$https://github.com/Expensify/App/issues/44185
PROPOSAL:

### Tests

- [ ] Check that the original bug is fixed, reproduction:
	- Spam the 'a' key, typing 'aaaaaa...'
	- Press send
	- Spam the 'b' key, typing 'bbbbbb...'
	- Expect that the composer shows first 'aaaaaa', then clears it, then it shows 'bbbbbb'
	- **NOTE:** In development mode there is still a severe lag on the JS thread, so it might be that the events are not happening in sync with the user action. It's not a goal of this PR to fix this performance issue (will create a new issue for that; the issue is not present in release mode interestingly)
- [ ] Check everything in regards to the composer:
	- add text
	- delete text
	- paste text
	- select text and replace it with pasting
	- insert emojis
	- try mentioning someone
	- use the :colon: format for adding emojis
	- check the composer is synchronised properly between tabs
	- All markdown commands work
	- Any other uses cases you can think of with the composer

### Offline tests

n/a

### QA Steps

Same as testing steps

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android_native.webm](https://github.com/Expensify/App/assets/16821682/284f2703-ad0f-4d85-9e91-cab2220de35c)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/16821682/75bfbc4c-5684-4351-b9c8-ceecd9cef4ec

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/16821682/4c7e6c71-eb20-4c42-8a7b-ec6237a05a11

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/16821682/55793806-dbbf-4126-bcbb-e9687d94cc0e

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/16821682/52a25bc4-6c35-4e15-9ce2-cb93275411c8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/16821682/7e9d91b3-fdbf-4efb-96be-11934621cc5a

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx [hunk 1]
```diff
@@ -46,7 +47,7 @@ import updateMultilineInputRange from '@libs/updateMultilineInputRange';
 import willBlurTextInputOnTapOutsideFunc from '@libs/willBlurTextInputOnTapOutside';
 import getCursorPosition from '@pages/home/report/ReportActionCompose/getCursorPosition';
 import getScrollPosition from '@pages/home/report/ReportActionCompose/getScrollPosition';
-import type {ComposerRef, SuggestionsRef} from '@pages/home/report/ReportActionCompose/ReportActionCompose';
+import type {SuggestionsRef} from '@pages/home/report/ReportActionCompose/ReportActionCompose';
```
Comment (hannojg) [inline src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx:50]: TODO to myself: revert this change to keep changes at a minimum

**Comment by hannojg** [comment]
Working on one last bug on web where deleting a multi-character glyph doesn't remove the whole glyph but just parts of it - otherwise I think the implementation is stable!

The web bug:


https://github.com/Expensify/App/assets/16821682/c65ff7e4-bd00-486c-8001-540f42c842b3

**Comment by hannojg** [comment]
**Update:** the web bug I [reported here](https://github.com/Expensify/App/pull/44285#issuecomment-2200524436) should be fixed as well now!

**Comment by hannojg** [comment]
A thanks for reporting, let me check!

**Comment by hannojg** [comment]
Can you reproduce it always on android (e.g. does it also not work if you add text before it?)

Do you see any issues in the metro bundler?

Asking because I can't reproduce the issue:

[Screen_recording_20240702_160528.webm](https://github.com/Expensify/App/assets/16821682/e122e186-d6a5-4ccb-9379-0f1cda545ef5)

**Comment by hannojg** [comment]
https://github.com/Expensify/App/assets/16821682/5601862e-9fb1-41c1-9612-d1559e7bb4e4

Currently investigating one bug where on web when selecting the whole text, deleting it, and then typing the full text reappears with the new text prefixed

**Comment by hannojg** [comment]
Okay, pushed a fix. Next bug I am looking into is when pasting the content is currently doubled:


https://github.com/Expensify/App/assets/16821682/8acfb872-cb6a-4bc6-8e2b-90846d0c37e4

**Comment by hannojg** [comment]
Hm, might be a regression that has been introduced here:

- https://github.com/Expensify/react-native-live-markdown/pull/369#issuecomment-2210752761

**Comment by Skalakid** [comment]
Hi @hannojg, we worked recently on fixing double pasting issue and we found a fix for it. I will create a PR that with the changes and live-markdown bump to the latest version. Is it okay for you?

**Comment by hannojg** [comment]
> Hi @hannojg, we worked recently on fixing double pasting issue and we found a fix for it. I will create a PR that with the changes and live-markdown bump to the latest version. Is it okay for you?

Sure! 😊

**Comment by Skalakid** [comment]
@hannojg Here is the [PR](https://github.com/Expensify/App/pull/44958) with the fix

**Comment by hannojg** [comment]
**TODO:** The rn-live-markdown PR has just been merged. I need to update to version `0.1.103` and remove the patch

**Comment by hannojg** [comment]
Added a PR here to fix one issue on the web:

- https://github.com/Expensify/react-native-live-markdown/pull/433

can you maybe have a look @Skalakid ?

**Comment by hannojg** [comment]
Found another bug in this PR:

- write aaaaa
- write bbbb
- press ctrl + z, text should be aaaaabbb
- write c
- Expected: aaaaabbbc
- Actual: aaaaabbbcb

**Comment by hannojg** [comment]
Opened this PR to address the bug:

- https://github.com/Expensify/react-native-live-markdown/pull/434

**Comment by hannojg** [comment]
I implemented a simpler solution over here:

- https://github.com/Expensify/App/pull/45613

**Comment by hannojg** [comment]
(we can reopen this PR in case the other PR wouldn't work for any reason)
